### PR TITLE
Implemented Dirichlet_Eta Derivative

### PR DESCRIPTION
**Pull request by vishusidana95**
@isuruf I have tried to implement the Dirichlet eta Derivative if you can review and suggest some changes.
This is in reference with the issue #814.

### CODE REVIEW

**Comment by vishusidana95** [comment]
@isuruf If you can please review it and suggest improvements?

File: symengine/derivative.cpp [hunk 1]
```diff
@@ -80,6 +80,25 @@ class DiffImplementation
...
     }
 
+    static RCP<const Basic> diff(const Dirichlet_eta &self,
+                                 const RCP<const Symbol> &x)
+    {
+        RCP<const Basic> m1;
+        m1 = zeta(self.get_arg())->diff(x);
+        // check https://en.wikipedia.org/wiki/Dirichlet_eta_function
+        // for the equation
+        if (eq(*self.get_arg(), *one)) {
```
Comment (isuruf) [inline symengine/derivative.cpp]: This doesn't make sense. `dirichlet_eta(one)->diff(x)` is zero, no matter what the symbol `x` is.

Comment (vishusidana95) [inline symengine/derivative.cpp]: as zeta(1)-->infinity and zeta'(1)--> -infinity the derivative is not defined at 1. So this is limit is calculated using Laurent expantion as you can read here:
http://math.stackexchange.com/questions/253549/problem-with-the-dirichlet-eta-function

or maybe I can remove this condition?

Comment (isuruf) [inline symengine/derivative.cpp]: This is not about value of `zeta(x).diff(x)` at `x = 1`, `zeta(1).diff(x) == 0`.
For eg. `cos(x).diff(x) == -sin(x)` and `cos(1).diff(x) == 0`

**Comment by vishusidana95** [comment]
@isuruf I removed that condition. Are there any other issue or improvement that you can suggest.

**Comment by isuruf** [comment]
Can you fix the git history? otherwise this is good to go

**Comment by isuruf** [comment]
Can you add some tests as well?

**Comment by vishusidana95** [comment]
@isuruf Thanks for the review. 
I have added Test Cases and fixed the git history as you said. Is there anything else you would suggest?

File: symengine/tests/basic/test_functions.cpp [hunk 2]
```diff
@@ -2304,6 +2309,28 @@ TEST_CASE("Dirichlet Eta: functions", "[functions]")
...
+    r1 = dirichlet_eta(i4);
+    r2 = div(mul(pow(pi, i4), i7), integer(720));
+    REQUIRE(eq(*r1, *r2));
+
+    r1 = dirichlet_eta(i6);
+    r2 = div(mul(pow(pi, i6), integer(31)), integer(30240));
+    REQUIRE(eq(*r1, *r2));
+
+    r1 = dirichlet_eta(i8);
+    r2 = div(mul(pow(pi, i8), integer(127)), integer(1209600));
```
Comment (isuruf) [inline symengine/tests/basic/test_functions.cpp]: All of the above tests `dirichlet_eta` value, not the derivative. Can you add more derivative tests?

Comment (vishusidana95) [inline symengine/tests/basic/test_functions.cpp]: @isuruf I have added tests for derivative can you review?

**Comment by vishusidana95** [comment]
@isuruf Can you review it??

**Comment by certik** [comment]
@vishusidana95 --- which derivative formula did you implement --- this one: https://en.wikipedia.org/wiki/Dirichlet_eta_function#Derivatives ?
```latex
\eta'(s) = 2^{1-s}\ln 2 \zeta(s)+(1-2^{1-s})\zeta'(s)
```
Here is the code that you implemented:
```c++
        return add(mul(mul(pow(i2, sub(one, self.get_arg())),
                           mul(log(i2), zeta(self.get_arg()))),
                       self.get_arg()->diff(x)),
                   mul(sub(one, pow(i2, sub(one, self.get_arg()))), m1));
```
and I am not sure that's the same thing. You are differentiating the argument `s` there, but rather it should be the derivative of `zeta` itself in the formula.

Or am I missing something? Also this seems to be only valid for `s != 1`. Note that SymPy doesn't evaluate the derivative either:
```
In [6]: dirichlet_eta(s).diff(s)
Out[6]: 
d                 
──(dirichlet_η(s))
ds                

In [7]: dirichlet_eta(s).diff(s).rewrite(zeta)
Out[7]: 
d ⎛⎛   -s + 1    ⎞     ⎞
──⎝⎝- 2       + 1⎠⋅ζ(s)⎠
ds                      
```

**Comment by vishusidana95** [comment]
@certik 
I declared m1 = derivative of zeta in the above line of code which I multiplied with other terms and I think when the argument will be one the derivative should give zero. But I am not sure exactly can you suggest something?

**Comment by ranjithkumar007** [comment]
@vishusidana95 Are you working on it ? If not i would like to wrap this up.
and derivative of dirichlet is not zero at 1..it is given by 
\eta '(1)=\ln(2)\gamma -\ln(2)^{2}/2   [reference](https://en.wikipedia.org/wiki/Dirichlet_eta_function#Derivatives)

**Comment by isuruf** [comment]
This was implemented in #1199

**Pull request closed without merging**